src/librustc_typeck/check/regionck.rs

```diff
@@ -759,6 +759,11 @@ impl<'a, 'gcx, 'tcx, 'v> Visitor<'v> for RegionCtxt<'a, 'gcx, 'tcx> {
             }
 
             hir::ExprRet(Some(ref ret_expr)) => {
+                if self.call_site_scope.is_none() {
+                    struct_span_err!(self.tcx.sess, expr.span, E0571,
+                                     "return statement cannot be out of a function scope").emit();
```

Comment:
I'm fine with banning it earlier - the return for constant values only makes sense in MIR anyway.